src/storage/storage.cpp

```diff
@@ -56,57 +49,19 @@ using QueryGraph = util::StaticGraph<contractor::QueryEdge::EdgeData>;
 
 Storage::Storage(StorageConfig config_) : config(std::move(config_)) {}
 
-struct RegionsLayout
-{
-    SharedDataType current_data_region;
-    boost::interprocess::named_sharable_mutex &current_region_mutex;
-    SharedDataType old_data_region;
-    boost::interprocess::named_sharable_mutex &old_region_mutex;
-};
-
-RegionsLayout getRegionsLayout(SharedBarriers &barriers)
-{
-    if (SharedMemory::RegionExists(CURRENT_REGION))
-    {
-        auto shared_region = makeSharedMemory(CURRENT_REGION);
-        const auto shared_timestamp =
-            static_cast<const SharedDataTimestamp *>(shared_region->Ptr());
-        if (shared_timestamp->region == REGION_1)
-        {
-            return RegionsLayout{
-                REGION_1, barriers.region_1_mutex, REGION_2, barriers.region_2_mutex};
-        }
-
-        BOOST_ASSERT(shared_timestamp->region == REGION_2);
-    }
-
-    return RegionsLayout{REGION_2, barriers.region_2_mutex, REGION_1, barriers.region_1_mutex};
-}
-
-Storage::ReturnCode Storage::Run(int max_wait)
+int Storage::Run(int max_wait)
 {
     BOOST_ASSERT_MSG(config.IsValid(), "Invalid storage config");
 
     util::LogPolicy::GetInstance().Unmute();
 
-    SharedBarriers barriers;
+    RetryLock datastore_lock(max_wait, "osrm-datastore-lock");
 
-    boost::interprocess::upgradable_lock<boost::interprocess::named_upgradable_mutex>
-        current_region_lock(barriers.current_region_mutex, boost::interprocess::defer_lock);
-    try
-    {
-        if (!current_region_lock.try_lock())
-        {
-            util::Log(logWARNING) << "A data update is in progress";
-            return ReturnCode::Error;
-        }
-    }
-    // hard unlock in case of any exception.
-    catch (boost::interprocess::lock_exception &)
+    if (!datastore_lock.TryLock())
     {
-        barriers.current_region_mutex.unlock_upgradable();
-        // make sure we exit here because this is bad
-        throw;
+        util::Log(logWARNING) << "Could not aquire current region lock after " << max_wait
+                              << " seconds. Claiming the lock by force.";
+        datastore_lock.ForceLock();
```

Comment:
This line can be a source of many problems, because it will make state of `CURRENT_REGION` undefined and anything could happen after executing `ForceLock`. I would suggest to use non-timed mutex, and add mutex removal to `CleanupSharedBarriers` to prevent making locked mutexes after signals.